New function: `hasPlayerSpawned`

TracerDS opened the pull request:
Returns `true` when player has been spawned by using `spawnPlayer`. `false` otherwise

TracerDS committed:
Added related functions

tederis commented on Client/mods/deathmatch/logic/CClientPed.cpp:
```diff
@@ -731,6 +731,8 @@ void CClientPed::Spawn(const CVector& vecPosition, float fRotation, unsigned sho
         SetFrozenWaitingForGroundToLoad(true);
         m_iLoadAllModelsCounter = 10;
+
+        static_cast<CClientPlayer&>(*this).SetPlayerSpawned(true);
```
I would avoid the expressions like this. Looks clumsy and violates the design principles. Have you thought about moving the entire method CClientPed::Spawn under the CClientPlayer class? Or

TracerDS replied:
Cant move it to the player. Players inherit functions from peds and `Spawn` manage the ped's being alive. If I move it to CClientPlayer then I will break CPed

tederis replied:
The only place from which this method is invoked is Packet_PlayerSpawn which on its own is dedicated for players only(not peds). It cannot break non-player peds.

TracerDS replied:
> The only place from which this method is invoked is Packet_PlayerSpawn which on its own is dedicated for players only(not peds). It cannot break non-player peds.

Maybe instead of moving the entire method, I'll move only `SetPlayerSpawned` there
```cpp
pPlayer->Spawn(vecPosition, fRotation, usPlayerModelID, ucInterior);
pPlayer->SetPlayerSpawned(true);
```

tederis replied:
That's better. But breaks the object's state coherency. Because someone in the future can invoke Spawn but forget to call SetPlayerSpawned and boom, object's state is invalid now.

TracerDS replied:
Okay, I'll move it to `CClientPlayer`

tederis replied:
I would suggest you to leave it as is. But after the merge create a new PR with the corresponding refactor.

TracerDS replied:
If it wont cause any issues then its better to do it now instead of waiting for it 'til merge

TracerDS committed:
Moved `CClientPed::Spawn` to `CClientPlayer` class

tederis commented on Server/mods/deathmatch/logic/CPlayer.h:
```diff
@@ -465,4 +468,6 @@ class CPlayer final : public CPed, public CClient
     ushort  m_usPrevDimension;
     SString m_strQuitReasonForLog;
+
+    bool m_bHasPlayerSpawned;
```
There already exists field CPed::m_bSpawned on server side. I know it behaves a bit different, but maybe there could be a way to avoid that collision. Because two fields with almost the same name may be confusing.

TracerDS replied:
> There already exists field CPed::m_bSpawned on server side. I know it behaves a bit different, but maybe there could be a way to avoid that collision. Because two fields with almost the same name may be confusing.

`m_bHasPlayerSpawned` will change to `false` when player dies. If I were to make these 2 variables into 1, I'd need to make a complete overhaul of the spawn system

tederis commented on Server/mods/deathmatch/logic/CMapManager.cpp:
```diff
@@ -592,6 +592,7 @@ void CMapManager::SpawnPlayer(CPlayer& Player, const CVector& vecPosition, float
     }
 
     // Update the player data
+    Player.SetPlayerSpawned(true);
```
I see that SetPlayerSpawned is called only ones. Should't there be a place where you call `SetPlayerSpawned(false)`?

TracerDS replied:
> I see that SetPlayerSpawned is called only ones. Should't there be a place where you call `SetPlayerSpawned(false)`?

You can't despawn a player

tederis replied:
> > I see that SetPlayerSpawned is called only ones. Should't there be a place where you call `SetPlayerSpawned(false)`?
> 
> You can't despawn a player

What will happened when player is died then?

TracerDS replied:
> What will happened when player is died then?

When player dies, `m_bHasPlayerSpawned` will change to `false` (indicating that the player has been despawned) but the player itself wont despawn, he will retain every property of a spawned player.
Its kinda confusing tbf. Dead players should be dead, not despawned.

TracerDS replied:
Actually, maybe I should just update `SetSpawned` to only work for spawning player and retain `SetDead` for killing/reviving players/peds? That would make more sense, no?

tederis replied:
> > What will happened when player is died then?
> 
> When player dies, `m_bHasPlayerSpawned` will change to `false` (indicating that the player has been despawned) but the player itself wont despawn, he will retain every property of a spawned player. Its kinda confusing tbf. Dead players should be dead, not despawned.

But how exactly the value of m_bHasPlayerSpawned can become false? There's no place where it's set to false in your code.

TracerDS replied:
> > > What will happened when player is died then?
> > 
> > 
> > When player dies, `m_bHasPlayerSpawned` will change to `false` (indicating that the player has been despawned) but the player itself wont despawn, he will retain every property of a spawned player. Its kinda confusing tbf. Dead players should be dead, not despawned.
> 
> But how exactly the value of m_bHasPlayerSpawned can become false? There's no place where it's set to false in your code.

Oops, sorry. I meant `m_bSpawned` (the original variable that changes when player dies)

`m_bHasPlayerSpawned` should only be set to false when player joins and leaves.
Since there is no direct way to despawn a player, it will never be set to false once a player has been spawned.

tederis replied:
> Actually, maybe I should just update `SetSpawned` to only work for spawning player and retain `SetDead` for killing/reviving players/peds? That would make more sense, no?

Yeah. I think `m_bSpawned` could be renamed to m_bDead (with the corresponding refactor, of course). When player connects m_bDead is false, when he's dead m_bDead is true (obviously), and when he's spawned after it m_bDead is false again. In this case we can avoid the collision and there will no any confusion.

TracerDS committed:
Updated code according to reviews

tederis commented on Server/mods/deathmatch/logic/CGame.cpp:
```diff
@@ -3735,8 +3734,8 @@ void CGame::Packet_VehicleTrailer(CVehicleTrailerPacket& Packet)
-        // Spawned?
-        if (pPlayer->IsSpawned())
+        // Alive?
+        if (!pPlayer->IsDead())
```
I see you has taken care of two call sites where IsSpawned was used. But there are more of them: 
![spawned](https://github.com/multitheftauto/mtasa-blue/assets/12121551/c095e3c4-65eb-411c-b094-219b5d546c14)

TracerDS replied:
I didnt know if I should replace them since these lines operate on alive peds

tederis replied:
That's important because you have changed the meaning of `m_bSpawned`.

tederis replied:
It's become complicated. What if you revert the last commit and just rename m_bHasPlayerSpawned with something like m_bAlive? Simple and no collisions.

TracerDS replied:
> That's important because you have changed the meaning of `m_bSpawned`.

I can change `IsSpawned` to `IsDead`. Based on the usage of the `SetSpawned` before the PR, `SetSpawned` is the same as `SetIsDead(not IsDead())`

TracerDS committed:
Changes `IsSpawned` to `IsDead` in some places

tederis commented on Server/mods/deathmatch/logic/CGame.cpp:
```diff
@@ -4296,9 +4295,6 @@ void CGame::PlayerCompleteConnect(CPlayer* pPlayer)
     pPlayer->Send(CPlayerConnectCompletePacket());
-
-    // The player is spawned when he's connected, just the Camera is not faded in/not targetting
-    pPlayer->SetSpawned(true);
```
You have replaced IsSpawned with !IsDead(). But at this line you can see that they are not interchangeable. Because m_bIsDead is only set when player is wasted. But m_bSpawned is true when player's connection is completed, not only when he's spawned via spawnPlayer.

TracerDS replied:
that line of code itself doesnt make sense. When player joins a server (lets say without any resources) then he only joined but he is not spawned. Only `spawnPlayer` will spawn a player. Currently, mta sets `m_bSpawned` to the opposite of `m_bIsDead` on every dying related event or function.

tederis replied:
Okay. But it should be noted that these changes will break the backward compatibility. Lets take a look at `setElementHealth` for example. Currently `setElementHealth` is valid even if it's called before `spawnPlayer`. After your changes it will not be true anymore.

tederis replied:
On the other hand, who the hell would want to use setElementHealth if player is not even spawned, right?

TracerDS replied:
> On the other hand, who the hell would want to use setElementHealth if player is not even spawned, right?

It also *might potentially* fix #472 but I really doubt it.

Backwards compatibility wouldn't really be a problem here. A quick refactor of `setElementHealth` to return `false` when the player is not spawned

Dutchman101 replied:
I've unresolved this review because as it stands, it looks like the backwards compatibility concern hasn't been tackled.
`A quick refactor of setElementHealth to return false when the player is not spawned` (@TracerDS)
Then please open a PR to do that, which will be a blocker for this to resolve and merge.

@tederis Do you have anything to add?

Dutchman101 replied:
@TracerDS I see you're updating the PR, but not acknownledging the above review.
Can you clarify why, because otherwise merge isn't gonna happen.

TracerDS replied:
> @TracerDS I see you're updating the PR, but not acknownledging the above review. Can you clarify why, because otherwise merge isn't gonna happen.

`setElementHealth` currently checks if player was spawned.
The value of `IsSpawned()` is the same as `!IsDead()`.
They were interchangeable. `m_bIsDead` is set to `false` in `CPed`'s constructor.
`m_bIsSpawned` is set to `true` **only** here.
So in current mta's code its always set to `true`.

In this PR you cannot set player's health when he is not connected to the server yet.
If it stayed the same then it might cause some issues.
Checking against `IsDead` instead fixes these issues and returns false when ped is dead/not spawned.
However, that would also mean that using `setElementHealth` on a dead ped will fail (return false). If that is a concern, I could remove the whole `if` statement here since its always set to `true` in previous code

FileEX commented on Server/mods/deathmatch/logic/luadefs/CLuaPlayerDefs.h:
```diff
@@ -47,6 +47,7 @@ class CLuaPlayerDefs : public CLuaDefs
     LUA_DECLARE(GetPlayerVersion);
     LUA_DECLARE(GetPlayerACInfo);
     static unsigned int GetPlayerScriptDebugLevel(CPlayer* const player);
+    static bool HasPlayerSpawned(CPlayer* player);
```
player can be const?

TracerDS replied:
~~When I marked any pointer `const`, ArgumentParser rejected these functions and threw a compile time error~~

Right, thanks

FileEX commented on Server/mods/deathmatch/logic/luadefs/CLuaPlayerDefs.cpp:
```diff
@@ -1459,6 +1460,11 @@ unsigned int CLuaPlayerDefs::GetPlayerScriptDebugLevel(CPlayer* const player)
     return player->GetScriptDebugLevel();
 }
 
+bool CLuaPlayerDefs::HasPlayerSpawned(CPlayer* player)
```
player can be const?

FileEX replied:
noexcept?

FileEX commented on Client/mods/deathmatch/logic/CClientPlayer.cpp:
```diff
@@ -205,6 +205,61 @@ bool CClientPlayer::IsOnMyTeam(CClientPlayer* pPlayer)
+    SetModel(usModel);
+
+    // Detach from any entities
+    AttachTo(NULL);
```
nullptr?

FileEX commented on Client/mods/deathmatch/logic/luadefs/CLuaPlayerDefs.cpp:
```diff
@@ -309,6 +310,13 @@ int CLuaPlayerDefs::GetPlayerWantedLevel(lua_State* luaVM)
 }
 
+
+bool CLuaPlayerDefs::HasPlayerSpawned(CClientPlayer* player)
```
noexcept?

TracerDS committed:
Merge github.com:TracerDS/mtasa-blue into 030624_NewFunction-isPlayerSpawned

TracerDS committed:
Update according to reviews

Also removed `hasPlayerSpawned` from client side so its only a server-side function

TracerDS committed:
Merge branch 'master' into 030624_NewFunction-isPlayerSpawned

TracerDS committed:
Update according to reviews

TracerDS committed:
Merge branch 'master' into 030624_NewFunction-isPlayerSpawned

TracerDS committed:
Merge branch 'master' into 030624_NewFunction-isPlayerSpawned

TracerDS committed:
Merge branch 'master' into 030624_NewFunction-isPlayerSpawned

Dutchman101 committed:
Merge branch 'master' into 030624_NewFunction-isPlayerSpawned

TracerDS committed:
Merge branch 'master' into 030624_NewFunction-isPlayerSpawned

TracerDS committed:
Merge branch 'master' into 030624_NewFunction-isPlayerSpawned

TracerDS committed:
Merge branch 'master' into 030624_NewFunction-isPlayerSpawned

TracerDS committed:
Merge branch 'master' into 030624_NewFunction-isPlayerSpawned

TracerDS committed:
Merge branch 'master' into 030624_NewFunction-isPlayerSpawned

TracerDS committed:
Merge branch 'master' into 030624_NewFunction-isPlayerSpawned

TracerDS committed:
Merge branch 'master' into 030624_NewFunction-isPlayerSpawned

TracerDS committed:
Merge branch 'master' into 030624_NewFunction-isPlayerSpawned

Fernando-A-Rocha reviewed (approved):
Good job

TracerDS committed:
Merge branch 'master' into 030624_NewFunction-isPlayerSpawned

TracerDS commented:
Im not interested in advancing this PR further anymore.
If someone wants to refresh it, feel free to do so.
